!htp #17384 - Generalised Security Directives to accept Bearer Token Authentication

pjan opened the pull request:
As per promise to @rkuhn, an improved shot to allow Bearer Token Authentication, taking into accounts the comments made by @jrudolph in #17384, post the #17502 merge. Since the `OAuth2BearerToken` is part of the akka codebase, I opted to add support for it within the http module rather than provide additional facilities to provide it as a third party library.
- Renamed `UserCredentials` to the more general `Credentials`, and its `userName` param to the more appropriate `identifier` in case it covers both `HttpBasicAuthentication` and `BearerTokenAuthentication`.
- Dropped `Basic` from the directives, as they now similarly accept `Authenticator`s that handle `BasicHttpCredentials` and `OAuth2BearerToken`
- Added and amended the tests for both scala and java side

pjan committed:
Generalised Authentication Directives to accept Bearer Tokens

Make type of credential authentication explicit

Cleanup

ktoso commented:
OK TO TEST

drewhk commented:
Ah, why is this PR tagged as "per", I thought this is persistence related :)

pjan commented:
@drewhk sorry for the confusion. Updated.

drewhk commented:
/cc @sirthias @jrudolph

jrudolph commented:
The API you suggest hides the fact which kind of authentication credentials the user provided, Basic or Bearer authentication. You could even say, that it treats Bearer authentication as a kind of Basic authentication where id and password have the same value.

While, as you said in your original PR, "HttpBasicAuthenticator and BearerTokenAuthenticator have almost exactly the same implementation", I wonder if we should logically treat them that way. Our policy is to be explicit about the HTTP primitives we use in the routing structure.

We are currently a bit short on capacity for driving this further, so, I'm a unsure about where to go with this for now.

drewhk commented:
Can we close this PR then and open a discussion ticket instead (if there is not yet one)?

sirthias commented:
I like the intent of providing full support for the `OAuth2BearerToken` as, as @pjan says, we already have a proper model for it but don't yet go all the way for supporting OAuth2 in the routing layer.

Regarding the current impl proposal I am, like @jrudolph, a bit hesitant to drop the `Basic` suffix from all the names. The `authenticateBasic` directive for example is quite clear on what it does. A fully general `authenticate` directive would have to be able to deal with *any* kind of authentication. Not just basic and OAuth. For example, there also is [Digest Authentication](https://en.wikipedia.org/wiki/Digest_access_authentication), which is quite different and, AFAICS, wouldn't fit the `authenticate` API proposed here.

How about we simply add `authenticateOAuth2` and support without touching the existing `authenticateBasic` structures? While that might appear to duplicate a few things I think it would actually improve clarity.

What do you think?

rkuhn commented:
I agree with keeping the API clean and simple and obvious, possible implementation code duplication can be sorted out independent from this primary concern.

pjan commented:
Thanks for the constructive feedback and concrete pointers @sirthias and @rkuhn. I'm happy to refactor it to keep the `authenticateBasic` untouched, and mirror it for `authenticateOAuth2` if that's what everyone agrees on.

I understand you're all very limited in bandwidth, and happily try to squeeze out some time to make this happen in return for all the great work you've been doing with Akka. Much much appreciated!

sirthias commented:
Great! Thanks a lot, @pjan!

drewhk commented:
We will cut RC4 sometime early next week, after that, we will not add new features until after 1.0

pjan commented:
Will hence flesh this out tomorrow. Thanks for the notice @drewhk

pjan commented:
Updated based on the comments. Tests fail on seemingly unrelated parts.

I'm blocking some time to make changes according to comments so this can be finished before RC4 as per notice of @drewhk

rkuhn commented:
Indeed, the failure is unrelated. Thanks for your commitment!

jypma commented:
If this is to be changed, I'd like to chime in with another use case. We use both OAuth2 and Basic auth. In our setup, we actually need to send the HTTP basic username and password to a backend system for verification. It'd currently a bit cumbersome that `UserCredential` doesn't allow one to get the password out directly; this basically means we have to manually extract the `HttpCredentials` instead, and pass them into the authenticator separately.

I understand the strong recommendation to do a secure `==` on the passwords, but could the password at least be made available as well? Maybe by a case (sub)class or extractor one can match on?

drewhk commented:
/cc @sirthias @jrudolph

drewhk commented:
@pjan this will be merged after we released 1.0, i.e. 1-2 weeks from now. Keep this PR open and we take care of the rest for you.

pjan commented:
@drewhk now that we're on 1.0, do I need to make any changes?

2m commented:
@pjan could you rebase this PR so it merges cleanly? Also please squash the last commit which only changes one line to the previous one.

Also please create an issue and then modify commit messages of this PR to follow the [contribution guidelines](https://github.com/akka/akka/blob/master/CONTRIBUTING.md#creating-commits-and-writing-commit-messages). This way we can follow what issues are fixed in what milestone. :)

drewhk commented:
@jrudolph is this LGTM for you?

rkuhn commented on akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/SecurityDirectivesSpec.scala:
```diff
@@ -11,26 +11,37 @@ import akka.http.scaladsl.model.headers._
     }
+    "reject requests with an OAuth2 Bearer Token Authorization header with 401" in {
+      Get() ~> Authorization(OAuth2BearerToken("myToken")) ~> Route.seal {
+        dontBasicAuth { echoComplete }
```
shouldn’t this use OAuth?

rkuhn commented:
@jrudolph Could you check this? Looks good to me, but I’d value experienced input on security directives …

@pjan Would you mind updating so that tests pass and the merge button is green? (and please also make one commit as foreseen in CONTRIBUTING.md) If you can’t make time I’ll pick this up next week and finish it off.

pjan commented:
@rkuhn @jrudolph took advantage of my one-day old son keeping me awake. Rebased, updated and squashed. Let me know if any other issues need to be addressed.

sirthias commented:
LGTM and: Congratulations, @pjan, amazing that you find the time for Akka PRs with such a fresh addition to the family! :)

ktoso commented:
LGTM, thanks a lot for working on this @pjan :-)
I think it should also get some docs, otherwise it's not as easy to discover. I'll open a ticket for documenting this.
